### DISCUSSION TITLE
Suggestion boolean as int like in C/C++

### DISCUSSION THREAD

**Discussion by DeafMan1983** [dotnet/runtime#96569] [Ideas] [answered]
Hello everyone, 

I would like to suggest in next version of language version should to pass `int` as `boolean` like in C/C++.

Example:
while statement:
```
while (done)
{
    ...
    done = 0;
}
```
or if else statement:
```
if (!SDL_CreateRendererAndWindow(...))
{
    ...
}
```
And more features - we would to request with `int` as `boolean`.

Because I already ported Gtk3 in C# without `g_object_ref()`
I use only `g_type_check_instance_cast()` because I have ported `gboolean` from `int` in C#

But I don't like to change to bool because I want to stay original porting 1:1.

example from gmarcos.h
```
#define FALSE (0)
#define TRUE (!FALSE)
```
But it seems `int` and it was typedef as `gboolean`.
How do I wish `boolean` should to have same to `int` like we wrote in C/C++ and C#.

I hope you understand my English. Sorry my bad English!

**Comment by tannergooding** [comment, marked as answer]
This would be a suggestion for https://github.com/dotnet/csharplang

It's worth noting, however, that this has been proposed before (having trouble finding the issue at the moment) and it is unlikely to ever happen. The decision to not allow integers to be implicitly treated as booleans was a very explicit design decision and avoids a number of issues that exist around such behavior. The behavior in C/C++ has led to various real world issues, CVEs, etc and modern languages have almost universally avoided that behavior accordingly.

**Comment by DeafMan1983** [comment]
Hello I know but I have made with ClangSharpPInvokeGenerator with Gtk3 . I think with class `Convert`:
From C source:
```
#define FALSE (0)
#define TRUE (!FALSE)
```
It becomes in C#
```
public static int FALSE = 0;
public static int TRUE = Convert.ToInt32(!Convert.ToBoolean(FALSE));
```
Is it a correct trick?

EDIT: It looks like in C/C++ Hahaha!
![image](https://github.com/dotnet/runtime/assets/57066679/a040bac1-cf69-4a51-ab3d-6887f27011ec)

It is simple to know like in boolean :D
![image](https://github.com/dotnet/runtime/assets/57066679/f728ec66-1bb7-4ae2-b027-cd155c4c5321)
That is easy like charm.

**Reply by tannergooding** [reply]
You would typically do `x != 0` to convert from `integer` to `boolean` and `x ? 1 : 0` to convert from `boolean` to `integer`

**Reply by Clockwork-Muse** [reply]
@DeafMan1983 - Note that the reason why @tannergooding is specifically checking `x != 0` is that in C/C++ `0` converts to `false`, and _everything else_ converts to `true`.  While constants or defines usually do explicitly use `1` for true, code doesn't necessarily always rely on that (most often when checking pointers, where `null` was traditionally `0` and then all potentially valid pointers have a value).

Whether or not such behavior is intended can be hard to judge in some cases, which is one of the reasons that it's best not to do this automatic conversion.
